Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -233,39 +274,70 @@ public final class ManifestLoader: ManifestLoaderProtocol {
-            return manifest
+        try temp_await{
+            self.load(packagePath: path,
+                      baseURL: baseURL,
+                      version: version,
+                      revision: revision,
+                      toolsVersion: toolsVersion,
+                      packageKind: packageKind,
+                      fileSystem: fileSystem,
+                      diagnostics: diagnostics,
+                      on: .global(),
+                      completion: $0)
         }
+    }
 
-        let manifest = try self.loadFile(
-            manifestPath: manifestPath,
-            baseURL: baseURL,
-            version: version,
-            revision: revision,
-            toolsVersion: toolsVersion,
-            packageIdentity: packageIdentity,
-            packageKind: packageKind,
-            fileSystem: fileSystem,
-            diagnostics: diagnostics
-        )
+    public func load(
+        packagePath path: AbsolutePath,
+        baseURL: String,
+        version: Version?,
+        revision: String?,
+        toolsVersion: ToolsVersion,
+        packageKind: PackageReference.Kind,
+        fileSystem: FileSystem? = nil,
+        diagnostics: DiagnosticsEngine? = nil,
+        on queue: DispatchQueue,
+        completion: @escaping (Result<Manifest, Error>) -> Void
+    ) {
+        do {
+            let fileSystem = fileSystem ?? localFileSystem
+            let manifestPath = try Manifest.path(atPackagePath: path, fileSystem: fileSystem)
+            let packageIdentity = PackageIdentity(url: baseURL)
+            let cacheKey = try ManifestCacheKey(packageIdentity: packageIdentity,
+                                                manifestPath: manifestPath,
+                                                toolsVersion: toolsVersion,
+                                                env: ProcessEnv.vars,
+                                                swiftpmVersion: Versioning.currentVersion.displayString,
+                                                fileSystem: fileSystem)
+
+            if self.useInMemoryCache, let manifest = self.memoryCache[cacheKey] {
+                // Inform the delegate (backwards compatibility)
+                self.delegate?.willLoad(manifest: manifestPath)
+                return completion(.success(manifest))
+            }
 
-        if self.useInMemoryCache {
-            self.memoryCache[cacheKey] = manifest
-        }
+            self.loadFile(manifestPath: manifestPath,
+                          baseURL: baseURL,
+                          version: version,
+                          revision: revision,
+                          toolsVersion: toolsVersion,
+                          packageIdentity: packageIdentity,
+                          packageKind: packageKind,
+                          fileSystem: fileSystem,
+                          diagnostics: diagnostics,
+                          on: queue) { result in
+
+                // cache positive results
```

Comment:
This is great that we're no longer caching failures.  That will help debugging especially (in real use I think most fixes involve changes to the manifest so the caching of failures might be less of a problem there today).